tapiriik/services/PolarFlow/polarflow.py

```diff
@@ -0,0 +1,270 @@
+        headers.update({"Authorization": "Bearer {}".format(serviceRecord.Authorization["OAuthToken"])})
+        return headers
+
+    def WebInit(self):
+        params = {'response_type':'code',
+                  'client_id': POLAR_CLIENT_ID,
+                  'redirect_uri': WEB_ROOT + reverse("oauth_return", kwargs={"service": "polarflow"})}
+        self.UserAuthorizationURL = "https://flow.polar.com/oauth2/authorization?" + urlencode(params)
+
+    def RetrieveAuthorizationToken(self, req, level):
+        code = req.GET.get("code")
+        params = {"grant_type": "authorization_code",
+                  "code": code,
+                  "redirect_uri": WEB_ROOT + reverse("oauth_return", kwargs={"service": "polarflow"})}
+
+        response = requests.post("https://polarremote.com/v2/oauth2/token", data=params, auth=HTTPBasicAuth(POLAR_CLIENT_ID, POLAR_CLIENT_SECRET))
+        data = response.json()
+
+        if response.status_code != 200:
+            raise APIException(data["error"])
+
+        authorizationData = {"OAuthToken": data["access_token"]}
+        userId = data["x_user_id"]
+
+        try:
+            self._register_user(data["access_token"])
+        except requests.exceptions.HTTPError as err:
+            # Error 409 Conflict means that the user has already been registered for this client.
+            # That error can be ignored
+            if err.response.status_code != 409:
+                raise APIException("Unable to link user", block=True, user_exception=UserException(UserExceptionType.Authorization, intervention_required=True))
+
+        return (userId, authorizationData)
+
+    def RevokeAuthorization(self, serviceRecord):
+        self._delete_user(serviceRecord)
+
+    def SubscribeToPartialSyncTrigger(self, serviceRecord):
+        # There is no per-user webhook subscription with Polar Flow.
+        serviceRecord.SetPartialSyncTriggerSubscriptionState(True)
+
+    def UnsubscribeFromPartialSyncTrigger(self, serviceRecord):
+        # As above.
+        serviceRecord.SetPartialSyncTriggerSubscriptionState(False)
+
+    def PollPartialSyncTrigger(self, multiple_index):
+        response = requests.get(self._api_endpoint + "/v3/notifications", auth=HTTPBasicAuth(POLAR_CLIENT_ID, POLAR_CLIENT_SECRET))
+
+        to_sync_ids = []
+        if response.status_code == 200:
+            for item in response.json()["available-user-data"]:
+                if item["data-type"] == "EXERCISE":
+                    to_sync_ids.append(item["user-id"])
+
+        return to_sync_ids
+
+    def DownloadActivityList(self, serviceRecord, exhaustive=False):
+        activities = []
+        exclusions = []
+        
+        transaction_url = self._create_transaction(serviceRecord)
+        if transaction_url:
+
+            res = requests.get(transaction_url, headers=self._api_headers(serviceRecord))
+            
+            if res.status_code == 200: # otherwise no new data, skip
```

Comment:
Would it make sense to at least check for 4xx/5xx codes here?

Reply:
Not sure. I think we should just skip case there is no data available because all possible errors are already been processed here: self._create_transaction(serviceRecord)